include/nbl/video/ILogicalDevice.h

```diff
@@ -539,19 +543,133 @@ class ILogicalDevice : public core::IReferenceCounted
+            if (m_defaultUploadBuffer.get()->needsManualFlushOrInvalidate())
+            {
+                IDriverMemoryAllocation::MappedMemoryRange flushRange(m_defaultUploadBuffer.get()->getBuffer()->getBoundMemory(), localOffset, subSize);
+                flushMappedMemoryRanges(1u, &flushRange);
+            }
+            // after we make sure writes are in GPU memory (visible to GPU) and not still in a cache, we can copy using the GPU to device-only memory
+            asset::SBufferCopy copy;
+            copy.srcOffset = localOffset;
+            copy.dstOffset = bufferRange.offset + uploadedSize;
+            copy.size = subSize;
+            cmdbuf->copyBuffer(m_defaultUploadBuffer.get()->getBuffer(), bufferRange.buffer.get(), 1u, &copy);
+            // this doesn't actually free the memory, the memory is queued up to be freed only after the GPU fence/event is signalled
+            /*event = */m_defaultUploadBuffer.get()->multi_free(1u, &localOffset, &subSize, core::smart_refctd_ptr<IGPUFence>(fence)); // can queue with a reset but not yet pending fence, just fine
+            uploadedSize += subSize;
+        }
+        //return event.deferred_function; // ????? wtf is deferred_function? why would i have an event here??
+    }
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUFence* fence, IGPUQueue* _queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        core::smart_refctd_ptr<IGPUCommandPool> pool = createCommandPool(_queue->getFamilyIndex(), IGPUCommandPool::ECF_RESET_COMMAND_BUFFER_BIT);
+        core::smart_refctd_ptr<IGPUCommandBuffer> cmdbuf;
+        createCommandBuffers(pool.get(), IGPUCommandBuffer::EL_PRIMARY, 1u, &cmdbuf);
+        assert(cmdbuf);
+        cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+        /*auto func = */updateBufferRangeViaStagingBuffer(cmdbuf.get(), fence, _queue, bufferRange, data);
+        cmdbuf->end();
+        IGPUQueue::SSubmitInfo submit;
+        submit.commandBufferCount = 1u;
+        auto* cmdbuf_raw = cmdbuf.get();
+        submit.commandBuffers = &cmdbuf_raw;
+        submit.signalSemaphoreCount = 0u;
+        submit.pSignalSemaphores = nullptr;
+        submit.waitSemaphoreCount = 0u;
+        submit.pWaitSemaphores = nullptr;
+        submit.pWaitDstStageMask = nullptr;
+        _queue->submit(1u, &submit, fence); // <threadsafeSubmit==true> means submit while locking an internal mutex
+        //func.optionalCmdBuffToDrop = std::move(cmdbuf); // ?????? wtf is optionalCmdBuffToDrop?
+    }
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUQueue* _queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        auto fence = this->createFence(static_cast<IGPUFence::E_CREATE_FLAGS>(0));
+        updateBufferRangeViaStagingBuffer(fence.get(), _queue, bufferRange, data);
+        auto* fenceptr = fence.get();
+        waitForFences(1u, &fenceptr, false, 9999999999ull);
```

Comment:
make the timeout greater XD, this is only 9 seconds (time unit is nanoseconds)

Reply:
didn't u say no gpu support more than 2s timeout anyway?

Reply:
by default, however windows has a registry key you can edit ;)